clarify missingness of point estimates; add examples for determining output_type_id format

zkamvar opened the pull request:
This modifies the description of `output_type_id` for point estimates to clearly state that the entry should be a missing value, either `NA` in R or `None` in Python. 

This also states that this applies to schemas prior to v4.0.0 (see https://github.com/hubverse-org/schemas/issues/109) and adds the python one-liner as suggested by @MKupperman in https://github.com/hubverse-org/hubDocs/issues/198.


The preview for this PR is at https://hubdocs--197.org.readthedocs.build/en/197/user-guide/model-output.html

zkamvar commented:
I would like to get another modeler perspective on this if it sounds clear: @trobacker

annakrystalli commented on docs/source/user-guide/model-output.md:
```diff
@@ -112,19 +116,20 @@ The following table provides more detail on how to configure the three "model ou
-
+[^missingno]: Point estimates don't have an `output_type_id` because you can only have one point estimate for each combination of task IDs.
+   However, because the `output_type_id` column is requrired, something has to go in this place, which is a missing value. 
+   This is encoded as [`NA_character_` in R](https://www.njtierney.com/post/2020/09/17/missing-flavour/) (which is why our schemas prior to 4.0.0 encoded these as `["NA"]`).
```
This is not always encoded as `NA_character_`. Output type ID columns may well be `double` if e.g. only `mean` and `quantile` output type IDs are being collected. The data type will be [determined by the overall config](https://hubverse.io/en/latest/user-guide/model-output.html#the-output-type-id-column-data-type). What we want to make clear is that `output_type_id` columns containing `NA` values must still adhere to the hub schema. Then we can provide tips on how to do this is different languages. E.g. in R you can assign `NA_real_` or  `NA_character_` or coerce to any other data type using the `as.*()` functions.

Also the `"NA"` in the config doesn't really have anything to do with a data type.  It is actually read in as the default logical:
``` r
jsonlite::fromJSON('{"required": ["NA"]}')$required |> typeof()
#> [1] "logical"
```
It's just the way to trick R to read an `NA` in as a bare `NA` is not allowed in JSON

We'd need to do e.g.:
``` r
jsonlite::fromJSON('{"required": ["NA_character_"]}')$required |> typeof()
#> [1] "character"
```
to get character but that's even more R specific so we avoided it.

annakrystalli replied:
Overall I feel this section is important enough and needs enough detailed discussion that it's own section is appropriate rather than a footnote

zkamvar replied:
> Overall I feel this section is important enough and needs enough detailed discussion that it's own section is appropriate rather than a footnote

Yeah I was realizing that nearly every cell in this table had a footnote, which meant that a new section is absolutely needed.

annakrystalli commented on docs/source/user-guide/model-output.md:
```diff
@@ -112,19 +116,20 @@ The following table provides more detail on how to configure the three "model ou
+[^missingno]: Point estimates don't have an `output_type_id` because you can only have one point estimate for each combination of task IDs.
+   However, because the `output_type_id` column is requrired, something has to go in this place, which is a missing value. 
+   This is encoded as [`NA_character_` in R](https://www.njtierney.com/post/2020/09/17/missing-flavour/) (which is why our schemas prior to 4.0.0 encoded these as `["NA"]`).
+   **If you use Python** to write parquet files, the `output_type_id` column should be an array of `None` and you will need to [explicitly cast the `output_type_id` column as a "string"](https://github.com/hubverse-org/hubValidations/issues/131#issuecomment-2427654006) (via Pandas):    
```
Not necessarily the entire column but any output type ID values associated with a point estimate output type. This may be the entire column if only point estimate output types are being submitted.

annakrystalli replied:
Also I don't feel a link to an issue comment is the best way to discuss this. As discussed above, I think a full section with more detailed advice would be more appropriate

annakrystalli commented on docs/source/user-guide/model-output.md:
```diff
@@ -112,19 +116,20 @@ The following table provides more detail on how to configure the three "model ou
 | ------ | ------ | ------ | 
-| `mean` | `"NA"`[^batman] (not used for mean predictions) | Numeric: the mean of the predictive distribution |
-| `median` | `"NA"` (not used for median predictions) | Numeric: the median of the predictive distribution |
+| `mean` | None[^missingno] (not used for mean predictions) | Numeric: the mean of the predictive distribution |
```
I would prefer at the very least something like `NA/None` (without any quotation marks) here with a prompt to more language specific details. That way `NA` is still easily recognisable as a correct value by R users. I believe `<NA>` also exists in python via `pd.NA`, is the closest thing to R's `NA` and I _think_ that's actually what @MKupperman used in his successful exploration (but I might be wrong).

In any case, I've renamed and moved #198 to here and have suggested we test out the various missingness options available in python so we have clarity about what works and what doesn't when we write this part of the documentation. Details in the issue.

annakrystalli commented on docs/source/user-guide/model-output.md:
```diff
@@ -112,19 +116,20 @@ The following table provides more detail on how to configure the three "model ou
-| `mean` | `"NA"`[^batman] (not used for mean predictions) | Numeric: the mean of the predictive distribution |
-| `median` | `"NA"` (not used for median predictions) | Numeric: the median of the predictive distribution |
+| `mean` | None[^missingno] (not used for mean predictions) | Numeric: the mean of the predictive distribution |
+| `median` | None (not used for median predictions) | Numeric: the median of the predictive distribution |
```
Same comment as above

micokoch reviewed (approved):
Anna made the technical comments, but from a readability angle, it looks good to me. It even made me realize I should uncapitalize a "parquet" which I had capitalized previously.

zkamvar commented:
I've updated the text according to Anna's suggestions and what I found in #198

elray1 commented:
I have not read all of this carefully.  But I noticed Anna's comment above, "I believe <NA> also exists in python via pd.NA, is the closest thing to R's NA and I think that's actually what @MKupperman used in his successful exploration (but I might be wrong)."

Just noting that [missing values in polars](https://docs.pola.rs/user-guide/expressions/missing-data/#null-and-nan-values) are best/most often/most generally represented using `null` (which is in practice set by specifying `None`), but `np.nan` may be used in the case of a column with floating point values.  Again, haven't read this but just wanted to quickly mention it in case it's relevant to anything here.

annakrystalli commented:
It might indeed be relevant @elray1.

What we are interested here is what missing values will serialise in an interoperable way that will be correctly interepreted as a missing value by our software when read in. We will definitely have `NA`s in our files as we will always have R users writing to `csv` files so either way polars will need to deal with that. I believe it can though as you would expect the R version of the polars package to understand `NA`s. There are also arguments when reading in files that allow us to specify null values on read (in both python and R).

Having said all that, it's of course best not to make any assumptions and actually test to ensure our suggestions don't cause any problems to desired downstream functionality. So @elray1, would you be able to add what a successful test of polars functionality requirements would look like here 🙏? https://github.com/hubverse-org/hubDocs/issues/198

zkamvar commented:
This PR is intended to clarify that `"NA"` in the schema actually means missing data and not the character `"NA"`. I've tested these assumptions in https://github.com/hubverse-org/hubDocs/issues/198#issuecomment-2430287818 and I think it satisfies this comment: 

> Having said all that, it's of course best not to make any assumptions and actually test to ensure our suggestions don't cause any problems to desired downstream functionality.

micokoch commented on docs/source/user-guide/model-output.md:
```diff
@@ -46,7 +46,11 @@ one-week-ahead incidence, but probabilities for the timing of a season peak:
 
-(formats-of-model-output)=
+[^batman]: `NA` (without quotes) indicates missingness in R, which is the expected `output_type_id` for a `mean` `output_type`. 
+  This is discussed in the [Output type table](#output-type-table)
```
```suggestion
  This is discussed in the [output type table](#output-type-table)
```

micokoch commented on docs/source/user-guide/model-output.md:
```diff
@@ -66,8 +70,10 @@ submission formats are _not mutually exclusive_; **hubs may choose between
       * Compatibility: Harder to work with; teams and people who want to work with files need to install additional libraries
 
+Examples of how to create these file formats in R and Python are listed below in
+[the Writing Model Output section](#writing-model-output).
```
```suggestion
[the writing model output section](#writing-model-output).
```

micokoch commented on docs/source/user-guide/model-output.md:
```diff
@@ -98,39 +104,135 @@ specified in the [section on usage of task ID variables](#task-id-use) when appr
+:::{note} 
+:name: output-type-caveats
+
+The model output type ids have different caveats depending on the `output_type`:
```
```suggestion
The model output type IDs have different caveats depending on the `output_type`:
```

micokoch commented on docs/source/user-guide/model-output.md:
```diff
@@ -98,39 +104,135 @@ specified in the [section on usage of task ID variables](#task-id-use) when appr
-  First, this provides a placeholder for the model output CSV file in the presence of other output types for validation.
-  The second reason is that we already use `null` to indicate the presence of an absence in the `required` and `optional` fields, and having this allows hubverse tools to treat point estimates without special handling. 
+ - `model_out_tbl` is the tabular output from your model formatted as specified
+   in [the Formats of Model Output section](#formats-of-model-output).
```
```suggestion
   in [the formats of model output section](#formats-of-model-output).
```

micokoch commented on docs/source/user-guide/model-output.md:
```diff
@@ -98,39 +104,135 @@ specified in the [section on usage of task ID variables](#task-id-use) when appr
+### Example: model output as CSV
 
-[^sample]: **Note on `sample` model output type**: Depending on the hub specification, samples with the same sample index (specified by the `output_type_id`) may be assumed to correspond to a single sample from a joint distribution across multiple levels of the task ID variables — further details are discussed below.
+To write to CSV, you would use the `write_csv()` from the readr package in R and
```
```suggestion
To write to CSV, you would use the `write_csv()` from the `readr` package in R and
```

micokoch reviewed (commented):
It looks very good, as far as I can tell. I made some small capitalization suggestions to be consistent with the rest of hubDocs.

zkamvar commented:
@annakrystalli, did I address your concerns?

annakrystalli commented on docs/source/user-guide/model-output.md:
```diff
@@ -85,37 +91,135 @@ Per hubverse convention, **there are two groups of columns providing metadata ab
-[^pmf]: **Note on `pmf` model output type**: Values are required to sum to 1 across all `output_type_id` values within each combination of values of task ID variables. This representation should only be used if the outcome variable is truly discrete; a CDF representation is preferred if the categories represent a binned discretization of an underlying continuous variable.
+ - `model_out_tbl` is the tabular output from your model formatted as specified
+   in [the formats of model output section](#formats-of-model-output).
+ - `path_to_hub` is the path to the hub cloned on your local computer
```
```suggestion
 - `hub_path` is the path to the hub cloned on your local computer
```
This is the argument name across the hubverse

annakrystalli commented on docs/source/user-guide/model-output.md:
```diff
@@ -85,37 +91,135 @@ Per hubverse convention, **there are two groups of columns providing metadata ab
+ - `model_out_tbl` is the tabular output from your model formatted as specified
+   in [the formats of model output section](#formats-of-model-output).
+ - `path_to_hub` is the path to the hub cloned on your local computer
+ - `model_name` is the file name of your model formatted as
```
```suggestion
 - `file_name` is the file name of your model formatted as
```

annakrystalli commented on docs/source/user-guide/model-output.md:
```diff
@@ -85,37 +91,135 @@ Per hubverse convention, **there are two groups of columns providing metadata ab
+In these examples, we are assuming the following variables already exist:
 
-[^pmf]: **Note on `pmf` model output type**: Values are required to sum to 1 across all `output_type_id` values within each combination of values of task ID variables. This representation should only be used if the outcome variable is truly discrete; a CDF representation is preferred if the categories represent a binned discretization of an underlying continuous variable.
+ - `model_out_tbl` is the tabular output from your model formatted as specified
```
This is not quite accurate. Model submissions do not generally contain `model_id` columns, When date is accessed from a hub via `coonect_hub()` it does. Only data accessed by a hub is a `model_out_tbl` class object.

zkamvar replied:
I'll go ahead and change it. I think when I wrote this, I did not fully realize that `model_out_tbl` was a reserved name.

annakrystalli commented on docs/source/user-guide/model-output.md:
```diff
@@ -85,37 +91,135 @@ Per hubverse convention, **there are two groups of columns providing metadata ab
+   in [the formats of model output section](#formats-of-model-output).
+ - `path_to_hub` is the path to the hub cloned on your local computer
+ - `model_name` is the file name of your model formatted as
+   `<round_id>-<model_name>.csv` (or parquet)
```
```suggestion
   `<round_id>-<model_id>.csv` (or parquet)
```
For consistency with hubverse terminology

annakrystalli commented on docs/source/user-guide/model-output.md:
```diff
@@ -85,37 +91,135 @@ Per hubverse convention, **there are two groups of columns providing metadata ab
+
+Writing to parquet is similar as writing to CSV, but with the caveat that you
+additionally need to ensure that the `output_type_id` column matches the
+[expected `output_type_id_datatype` property of the schema](#output-type-id-datatype).
```
That's only if the hub has specified an `output_type_id_datatype`. It might be just auto detected. See https://hubverse.io/en/latest/user-guide/model-output.html#the-output-type-id-column-data-type

This also might need to be performed regardless of the format being written out?. E.g is a column written out from python that contains `NA`s or `None` plus integers still written out as integers?

zkamvar replied:
> E.g is a column written out from python that contains NAs or None plus integers still read into R as integers?

Oh good catch! I had updated https://github.com/hubverse-org/hubDocs/issues/198#issuecomment-2430287818 with my findings, but I forgot to update here (in which pandas needs to convert to `Int64` to allow for integer `None` values). 

**The answer to your question is yes**. Even if a parquet file is written with python (either polars or pandas), the type stays with the parquet file and a `null` value inside an integer column is represented as an integer `NA` in R. 

The only thing that seems to have a problem with integer `null` values is pandas. By default, if it reads in a parquet file that has `null` values in an integer column, that column will be coerced to a float (unless the `dtype_backend` is set to `pyarrow`).

annakrystalli replied:
I think we should make it clear that `NA`s might also pop up in task IDs, not just the `output_type_id` column, for example [when a certain task ID is relevant to one modeling task but not another](https://hubverse.io/en/latest/quickstart-hub-admin/tasks-config.html#required-and-optional-elements) (also discussed in https://github.com/hubverse-org/hubValidations/issues/140)

annakrystalli commented on docs/source/user-guide/model-output.md:
```diff
@@ -85,37 +91,135 @@ Per hubverse convention, **there are two groups of columns providing metadata ab
+additionally need to ensure that the `output_type_id` column matches the
+[expected `output_type_id_datatype` property of the schema](#output-type-id-datatype).
+In practice, you will need to know whether or not the expected data type is a
+**string/character** or a **float/numeric**.
```
Related to this PR but not expected to be resolved by it, it's clear submitting teams need to have an idea of the hub schema, especally of the `output_type_id` column datatype. While `hubData::create_hub_schema()` and even `hubValidations::submission_tbl()` can provide this info in R, I'm wondering if there's a better way to communicate this infor to submitting teams more broadly?

zkamvar replied:
I think this is exactly the information I needed to know! I was initially confused by the use of `hubData::create_hub_schema()` since I thought it was related to our schemas repository. I had not used these functions and I've learned that some modelers were unaware that hubValidations even existed. At the moment, I think most people are scanning the schema to figure out how to format their data, but the table would likely be helpful and I think it's worthwhile to add into this section. 

As an aside, I wonder if we could add in a function somewhere (maybe in `hubData`) that just provides the expected `output_type_id` datatype given the schema like:

```r
output_type_id_datatype()
#> "numeric"
```

zkamvar replied:
I went ahead and iterated on this, rearranging some things and creating examples.

zkamvar commented on docs/source/user-guide/model-output.md:
````diff
@@ -85,37 +91,135 @@ Per hubverse convention, **there are two groups of columns providing metadata ab
 
+#### Writing CSV with R
+
+```r
+library("fs")
+library("readr")
+# ... generate model data ...
+outfile <- path(path_to_hub, "model-output", "team1-modelA", model_name)
+write_csv(model_out_tbl, outfile)
+```
+
+#### Writing CSV with Python
+
+```python
+import pandas as pd
+import os.path
+# ... generate model data ...
+outfile = os.path.join(path_to_hub, "model-output", "team1-modelA", model_name)
+model_out_tbl.to_csv(outfile, index = False, na_rep = "NA")
+```
````
````suggestion
```r
library("fs")
library("readr")
# ... generate model data ...
outfile <- path(hub_path, "model-output", "team1-modelA", model_id)
write_csv(model_out_tbl, outfile)
```

#### Writing CSV with Python

```python
import pandas as pd
import os.path
# ... generate model data ...
outfile = os.path.join(hub_path, "model-output", "team1-modelA", model_id)
model_out_tbl.to_csv(outfile, index = False, na_rep = "NA")
```
````

zkamvar commented on docs/source/user-guide/model-output.md:
````diff
@@ -85,37 +91,135 @@ Per hubverse convention, **there are two groups of columns providing metadata ab
+
+
+#### Writing parquet with R
+
+```r
+library("fs")
+library("arrow")
+# ... generate model data ...
+outfile <- path(path_to_hub, "model-output", "team1-modelA", model_name)
+model_out_tbl$output_type_id <- as.character(model_out_tbl$output_type_id) # or as.numeric()
+arrow::write_parquet(model_out_tbl, outfile)
+```
+
+
+#### Writing parquet with Python
+
+```python
+import pandas as pd
+import os.path
+# ... generate model data ...
+outfile = os.path.join(path_to_hub, "model-output", "team1-modelA", model_name)
+model_out_tbl["output_type_id"] = model_out_tbl["output_type_id"].astype("string") # or "float"
+model_out_tbl.to_parquet(outfile)
+```
````
````suggestion

```r
library("fs")
library("arrow")
# ... generate model data ...
outfile <- path(hub_path, "model-output", "team1-modelA", model_id)
model_out_tbl$output_type_id <- as.character(model_out_tbl$output_type_id) # or as.numeric()
arrow::write_parquet(model_out_tbl, outfile)
```


#### Writing parquet with Python

```python
import pandas as pd
import os.path
# ... generate model data ...
outfile = os.path.join(hub_path, "model-output", "team1-modelA", model_id)
model_out_tbl["output_type_id"] = model_out_tbl["output_type_id"].astype("string") # or "float"
model_out_tbl.to_parquet(outfile)
```
````

annakrystalli commented on docs/source/user-guide/model-output.md:
```diff
@@ -78,44 +86,207 @@ Much of the material in this section has been excerpted or adapted from the [hub
-Per hubverse convention, **there are two groups of columns providing metadata about the prediction**[^model-id], followed by **a value column with the actual output**. Each group of columns serves a specific purpose: (1) the **"task ID"** columns provide details about what is being predicted, and (2) the two **"model output representation"** columns specify the type of prediction and identifying  ...
+Per hubverse convention, **there are two groups of columns providing metadata about the prediction**[^model-id], followed by **a value column with the actual output**. Each group of columns serves a specific purpose: (1) the **"task ID"** columns provide details about what is being predicted, and (2) the two **"model output representation"** columns specify the type of prediction and identifying  ...
 
 [^model-id]: When using models for downstream analysis with the [`collect_hub()` function](https://hubverse-org.github.io/hubData/reference/collect_hub.html) in the `hubData` package, one more column called `model_id` is prepended added that identifies the model from its filename. 
```
```suggestion
[^model-id]: When using models for downstream analysis with the [`collect_hub()` function](https://hubverse-org.github.io/hubData/reference/collect_hub.html) in the `hubData` package, one more column called `model_id` is prepended that identifies the model from its filename. 
```

annakrystalli commented on docs/source/user-guide/model-output.md:
```diff
@@ -78,44 +86,207 @@ Much of the material in this section has been excerpted or adapted from the [hub
 
 [^model-id]: When using models for downstream analysis with the [`collect_hub()` function](https://hubverse-org.github.io/hubData/reference/collect_hub.html) in the `hubData` package, one more column called `model_id` is prepended added that identifies the model from its filename. 
 
+The model output follows the specification of the `tasks.json` configuration
```
While the following section is useful and detailed, it feels like it should be more of an aside towards the end of the discussion (i.e. it flows after the section where we tell folks that - btw your missing values need to be typed!). It is also inevitably very R centric when in some ways we're trying to be a little more cross-platform in this section. 

I think it would actually be better to have a way hubs can include schema information in a cross-platform way in their documentation, like a function that prints a table to their README. As I mentioned though this is a separate issue and more generally related to, in my opinion, work we need to do make metadata associated with hub data more visible/accessible.

So for now, I think we should keep this section as it's got lots of useful info but maybe move it down and link it to the mention of typed missing values.

zkamvar replied:
Yeah, I wasn't super happy with the placement of this and I agree with your sentiment. This is further evidence I need to split the documentation into the Diátaxis framework. 

> I think it would actually be better to have a way hubs can include schema information in a cross-platform way in their documentation, like a function that prints a table to their README. As I mentioned though this is a separate issue and more generally related to, in my opinion, work we need to do make metadata associated with hub data more visible/accessible.

Y'know, I wouldn't be surprised if this could be achieved with some vanilla in-browser JS.

annakrystalli replied:
Oooh that could be cool! Maybe we should open an issue somewhere to track this? (Not sure where the best place would be...)

zkamvar replied:
I opened it in the site builder dashboard repository.

annakrystalli commented on docs/source/user-guide/model-output.md:
````diff
@@ -78,44 +86,207 @@ Much of the material in this section has been excerpted or adapted from the [hub
+
+```r
+# read the configuration file and get the latest round
+config_tasks <- hubData::read_config_file(tasks_path)
````
You could use `hubData::read_config(hub_path)` instead and just provide the path to the hub

zkamvar replied:
I was thinking about @trobacker's comment that he usually models outside of the context of a hub, but I can certainly change this.

annakrystalli replied:
They still need access to the hub to get to the config file though? The only difference between the two functions is that one takes the full path to the config file and the other the path to a hub and then builds the config path relative to the root of the hub. It also defaults to the `tasks.json` which is the config most commonly loaded.

annakrystalli commented on docs/source/user-guide/model-output.md:
```diff
@@ -78,44 +86,207 @@ Much of the material in this section has been excerpted or adapted from the [hub
-[^pmf]: **Note on `pmf` model output type**: Values are required to sum to 1 across all `output_type_id` values within each combination of values of task ID variables. This representation should only be used if the outcome variable is truly discrete; a CDF representation is preferred if the categories represent a binned discretization of an underlying continuous variable.
+ - `model_out` is the tabular output from your model formatted as specified
+   in [the formats of model output section](#formats-of-model-output).
+ - `hub_path` is the path to the hub cloned on your local computer
```
```suggestion
 - `hub_path` is the path to the hub cloned on your local computer
 - `model_id` is the combination of `<team_abbr>-<model_abbr>` 
```

annakrystalli commented on docs/source/user-guide/model-output.md:
```diff
@@ -78,44 +86,207 @@ Much of the material in this section has been excerpted or adapted from the [hub
+library("fs")
+library("readr")
+# ... generate model data ...
+outfile <- path(hub_path, "model-output", "team1-modelA", model_id)
```
```suggestion
outfile <- fs::path(hub_path, "model-output", model_id, file_name)
```

annakrystalli commented on docs/source/user-guide/model-output.md:
````diff
@@ -78,44 +86,207 @@ Much of the material in this section has been excerpted or adapted from the [hub
+#### Writing CSV with R
+
+```r
+library("fs")
````
```suggestion
```

annakrystalli commented on docs/source/user-guide/model-output.md:
````diff
@@ -78,44 +86,207 @@ Much of the material in this section has been excerpted or adapted from the [hub
+
+```r
+library("fs")
+library("readr")
````
```suggestion
```
For conciseness

annakrystalli commented on docs/source/user-guide/model-output.md:
```diff
@@ -78,44 +86,207 @@ Much of the material in this section has been excerpted or adapted from the [hub
+library("readr")
+# ... generate model data ...
+outfile <- path(hub_path, "model-output", "team1-modelA", model_id)
+write_csv(model_out, outfile)
```
```suggestion
readr::write_csv(model_out, outfile)
```

annakrystalli commented on docs/source/user-guide/model-output.md:
```diff
@@ -78,44 +86,207 @@ Much of the material in this section has been excerpted or adapted from the [hub
+library("fs")
+library("arrow")
+# ... generate model data ...
+outfile <- path(hub_path, "model-output", "team1-modelA", model_id)
```
```suggestion
outfile <- fs::path(hub_path, "model-output", model_id, file_name)
```

annakrystalli commented on docs/source/user-guide/model-output.md:
````diff
@@ -78,44 +86,207 @@ Much of the material in this section has been excerpted or adapted from the [hub
+#### Writing parquet with R
+
+```r
+library("fs")
````
```suggestion
```

annakrystalli commented on docs/source/user-guide/model-output.md:
````diff
@@ -78,44 +86,207 @@ Much of the material in this section has been excerpted or adapted from the [hub
+
+```r
+library("fs")
+library("arrow")
````
```suggestion
```

annakrystalli commented on docs/source/user-guide/model-output.md:
```diff
@@ -78,44 +86,207 @@ Much of the material in this section has been excerpted or adapted from the [hub
-  First, this provides a placeholder for the model output CSV file in the presence of other output types for validation.
-  The second reason is that we already use `null` to indicate the presence of an absence in the `required` and `optional` fields, and having this allows hubverse tools to treat point estimates without special handling. 
+When submitting model output to a hub, it should be placed in a folder with the
+name of your model in the model outputs folder specified by the hub
```
```suggestion
name of your `model_id` in the model outputs folder specified by the hub
```
`model_id` is not the same as model name. It's the combination of `team_abbr` and `model_abbr`. This is discussed at the top of the page and might be a good idea to link to it.

zkamvar replied:
Oh good point! I did not realize this!

annakrystalli reviewed (changes requested):
Getting there from my perspective with some really useful and detailed additions!

Made a few corrections and a suggestion for moving a section and cleaning up some possible misconceptions.

zkamvar commented on docs/source/user-guide/model-output.md:
````diff
@@ -78,44 +86,207 @@ Much of the material in this section has been excerpted or adapted from the [hub
+
+```r
+# read the configuration file and get the latest round
+config_tasks <- hubData::read_config_file(tasks_path)
````
🙈

bsweger commented on docs/source/user-guide/model-output.md:
```diff
@@ -78,44 +86,204 @@ Much of the material in this section has been excerpted or adapted from the [hub
+point estimate for each combination of task IDs. However, because the
+`output_type_id` column is required, something has to go in this place, which
+is a missing value. This is encoded as [`NA` in
+R](https://www.njtierney.com/post/2020/09/17/missing-flavour/) (which is why
+our schemas prior to 4.0.0 encoded these as `["NA"]`) and `None` in Python. See
```
```suggestion
R](https://www.njtierney.com/post/2020/09/17/missing-flavour/) and `None` in Python. See
```
I think this reads better without the aside about schema versions...the important thing is to convey what the `output_type_id` column will look like for point estimate tasks

zkamvar replied:
totes agree

zkamvar commented on docs/source/user-guide/model-output.md:
```diff
@@ -78,44 +86,207 @@ Much of the material in this section has been excerpted or adapted from the [hub
+import os.path
+# ... generate model data ...
+outfile = os.path.join(hub_path, "model-output", "team1-modelA", model_id)
+model_out.to_csv(outfile, index = False, na_rep = "NA")
```
Yeah it's nuanced and annoying. I should probably just leave it out and allow NA's to be blanks. R should be able to handle them.

zkamvar replied:
I've looked at this again and I've removed the `na_rep` argument to simplify the example and it works.

bsweger commented on docs/source/user-guide/model-output.md:
```diff
@@ -78,44 +86,204 @@ Much of the material in this section has been excerpted or adapted from the [hub
+(example-parquet)=
+### Example: model output as parquet
+
+Writing to parquet is similar as writing to CSV, but with the caveat that you
+additionally need to ensure that the `output_type_id` column matches the
+[expected `output_type_id_datatype` property of the schema](#output-type-id-datatype).
```
```suggestion
Unlike a CSV, a parquet files contain embedded information about the data types of its
columns. Therefore, when writing model output files as parquet, it's critical that you
first ensure the data type of the `output_type_id` column matches the
[expected `output_type_id_datatype` property of the schema](#output-type-id-datatype).

If the data types of the model output parquet file don't match the hub's schema, the
submission will not validate.
```

Tried to surface the reason for the difference between csv and parquet, without getting too in-the-weeds. Also think it might be helpful to clarify why it's important to mind the parquet column types (maybe that would be a good message for one of those fancy warning boxes?)

zkamvar replied:
I think this is much better wording!

bsweger commented on docs/source/user-guide/model-output.md:
```diff
@@ -78,44 +86,204 @@ Much of the material in this section has been excerpted or adapted from the [hub
+import os.path
+# ... generate model data ...
+outfile = os.path.join(hub_path, "model-output", "team1-modelA", model_id)
+model_out["output_type_id"] = model_out["output_type_id"].astype("string") # or "float", or "Int64"
```
```suggestion
# update the output_type_id data type to match the hub's schema
model_out["output_type_id"] = model_out["output_type_id"].astype("string") # or "float", or "Int64"
```

bsweger commented on docs/source/user-guide/model-output.md:
````diff
@@ -78,44 +86,204 @@ Much of the material in this section has been excerpted or adapted from the [hub
+```r
+# ... generate model data ...
+outfile <- fs::path(hub_path, "model-output", model_id, file_name)
+model_out$output_type_id <- as.character(model_out$output_type_id) # or as.numeric(), or as.integer()
````
```suggestion
# update the output_type_id data type to match the hub's schema
model_out$output_type_id <- as.character(model_out$output_type_id) # or as.numeric(), or as.integer()
```

bsweger reviewed (commented):
Thanks for taking the time to make sure I'm understanding the problem space correctly! Based on that, I made a few more wording suggestions for your consideration (feel free to disregard)

zkamvar commented on docs/source/user-guide/model-output.md:
```diff
@@ -46,7 +46,11 @@ one-week-ahead incidence, but probabilities for the timing of a season peak:
 :::
 
-(formats-of-model-output)=
+[^batman]: `NA` (without quotes) indicates missingness in R, which is the expected `output_type_id` for a `mean` `output_type`. 
```
This might lead people to encode `None` in the CSV itself, which would be interpreted as a character. What about this: 

```suggestion
[^batman]: The `output_type_id` for point estimates (e.g. `mean`) is not applicable. To reflect this, we need to signal that this is a missing value. While a blank cell could indicate missingness, we chose `NA` because it's a clear indicator that it's a presence of an absence and not a mistake. In R, missing values are encoded as `NA` and in Python, they are encoded as `None`.
```

bsweger replied:
> This might lead people to encode `None` in the CSV itself, which would be interpreted as a character. What about this:

I take your point! But I find this confusing (perhaps because it's R-centric). Maybe less is more?
```suggestion
[^batman]: The `output_type_id` for point estimates (e.g. `mean`) is not applicable. To
reflect this, we need to signal that this is a missing value. In R, missing values are
encoded as `NA`, and in Python, they are encoded as `None`.
```

zkamvar replied:
Sold!

LucieContamin reviewed (commented):
Thank you for the update, it looks all clear to me, I just made some minor comments.

zkamvar commented on docs/source/user-guide/model-output.md:
```diff
@@ -78,44 +86,204 @@ Much of the material in this section has been excerpted or adapted from the [hub
+
+### Arrow Schema
+
+**The hubverse package `hubData()` has functionality that will generate an
```
good catch!

zkamvar commented on docs/source/user-guide/model-output.md:
```diff
@@ -78,44 +86,207 @@ Much of the material in this section has been excerpted or adapted from the [hub
+rounds <- hubData::get_round_ids(config_tasks)
+this_round <- rounds[length(rounds)]
+# create the submission template (this may take some time if your submission uses samples)
+tmpl <- hubValidations::submission_tmpl(config_tasks, round_id = this_round)
```
Thank you! The assertion on `hub_con` was actually a failure from my earlier code where I did not name the argument.

zkamvar commented:
Thank you @annakrystalli, @bsweger, @LucieContamin, and @micokoch for your comments!

As per @annakrystalli and @LucieContamin's suggestions, I have moved the documentation about the template generation and arrow schema closer to the examples. Specifically, I added the section about the arrow schema _inside_ the example so that users do not have to look far to know what the arrow schema would look like.

As per @bsweger's suggestions (and https://github.com/hubverse-org/hubDocs/issues/198#issuecomment-2448127804), I have attempted to focus the discussion for the modelers on the column type.

annakrystalli commented on docs/source/user-guide/model-output.md:
```diff
@@ -78,44 +88,229 @@ Much of the material in this section has been excerpted or adapted from the [hub
+#### Arrow Schema
 
-[^pmf]: **Note on `pmf` model output type**: Values are required to sum to 1 across all `output_type_id` values within each combination of values of task ID variables. This representation should only be used if the outcome variable is truly discrete; a CDF representation is preferred if the categories represent a binned discretization of an underlying continuous variable.
+**The hubverse packages `hubData()` and `hubUtils()` have functionality that will generate an
```
```suggestion
**The hubverse packages `hubData` and `hubUtils` have functionality that will generate an
```

annakrystalli commented on docs/source/user-guide/model-output.md:
```diff
@@ -78,44 +88,229 @@ Much of the material in this section has been excerpted or adapted from the [hub
+(submission-template)=
+### Submission Template
+
+**The hubverse package `hubValidations()` has functionality
```
```suggestion
**The hubverse package `hubValidations` has functionality
```
As its a pkg name and not a function name, we don't need parentheses

LucieContamin replied:
I nearly suggested it but was not sure I was correct, only function need parentheses, no? (Just to be sure I have the correct standard)

annakrystalli commented on docs/source/user-guide/model-output.md:
````diff
@@ -78,44 +88,229 @@ Much of the material in this section has been excerpted or adapted from the [hub
+value: int32
+model_id: string
+```
+
````
```suggestion

If you have a data frame of model output data (e.g. `model_out`) and want to coerce the column data types to the hub schema, you can also use [`hubData::coerce_to_hub_schema()`](https://hubverse-org.github.io/hubData/reference/coerce_to_hub_schema.html), e.g. 
```r
hubData::coerce_to_hub_schema(model_out, config_tasks)
```

```

annakrystalli reviewed (approved):
I made a couple of minor last minute corrections and added a suggested note about another useful function, `hubData::coerce_to_hub_schema()`.

Other than that, all my comments have been resolved. Thanks @zkamvar !

annakrystalli commented:
FYI as a response to functions being relevant to modelers appearing across too many packages, I've made a proposal to re-export them to `hubValidations` here: https://github.com/hubverse-org/hubValidations/issues/149

Input welcome

zkamvar commented on docs/source/user-guide/model-output.md:
```diff
@@ -78,44 +88,229 @@ Much of the material in this section has been excerpted or adapted from the [hub
+outfile <- fs::path(hub_path, "model-output", model_id, file_name)
 
+# update the output_type_id data type to match the hub's schema
+model_out$output_type_id <- as.character(model_out$output_type_id) # or as.numeric(), or as.integer()
```
oh good point!

zkamvar commented on docs/source/user-guide/model-output.md:
```diff
@@ -78,44 +88,229 @@ Much of the material in this section has been excerpted or adapted from the [hub
+outfile = os.path.join(hub_path, "model-output", model_id, file_name)
+
+# update the output_type_id data type to match the hub's schema
+model_out["output_type_id"] = model_out["output_type_id"].astype("string") # or "float", or "Int64"
```
thank you!

zkamvar commented:
Thank you all for sticking with me through this!
